### PR TITLE
Rule S1264: A 'while' loop should be used instead of a 'for' loop

### LINKED ISSUE #1172: Rule S1264: A "while" loop should be used instead of a "for" loop
Implements [RSPEC-1264](https://jira.sonarsource.com/browse/RSPEC-1264)

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #1172

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseWhileLoopInstead.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var forStatement = (ForStatementSyntax)c.Node;
+
+                    if (forStatement.Declaration == null &&
+                        forStatement.Incrementors.Count == 0)
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseWhileLoopInstead.cs]: It looks a odd that empty declaration is null and empty incrementor is a non-null empty collection...
Could `Incrementors` be null?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UseWhileLoopInstead.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            var x = 5;
+            for (; x < 10;) // Noncompliant {{Replace this 'for' loop with a 'while' loop.}}
+//          ^^^
+            {
+            }
+
+            for (; ; ) // Noncompliant
+            {
+
+            }
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UseWhileLoopInstead.cs]: Other compliant cases:
* just the declaration is null
* just the incrementor is null

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseWhileLoopInstead.cs (hunk 1 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseWhileLoopInstead.cs]: Haven't found a way but as I told you we should go defensive I am going to handle the case anyway.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UseWhileLoopInstead.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            {
+            }
+
+            var x = 5;
+            for (; x < 10;) // Noncompliant {{Replace this 'for' loop with a 'while' loop.}}
+//          ^^^
+            {
+            }
+
+            for (; ; ) // Noncompliant
```
Comment (valhristov) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UseWhileLoopInstead.cs]: Not sure if `while(true) { }` is much better than `for (;;) { }`... IMO this syntax is a matter of personal preference and could be controversial, e.g. I would probably not raise here.

Comment (Evangelink) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/UseWhileLoopInstead.cs]: That's the rule example.

**Pull request merged**